### PR TITLE
video: wayland: Use wp-viewporter for fullscreen with fractional scaling

### PR DESCRIPTION
**Pull request by Kontrabant**
Desktops with fractional scaling need special handling to avoid significant overdraw as desktop-sized windows can have backbuffers significantly larger than the native resolution, which then have to be downscaled for final output.  This uses the wp_viewporter protocol, when available, to allow for the use of the native display resolution in fullscreen applications and, potentially, direct scanout by compositors.

Desktops with fractional scaling that don't support viewports are still stuck rendering to oversized backbuffers, but there's nothing we can do about this (do any even exist?)

Needless to say, this needs testing on as many desktops as possible.  It works well on GNOME 41 with Vulkan, GL and the software blitter, but I haven't had a chance to test it on desktops like KDE or Sway.  My main concern is if setting the window to larger than the desktop will have any adverse effects.  This wasn't necessary for Vulkan and GL, but the SDL software blitter, for instance, still uses the window size to set the surface dimensions and having mismatched window and backbuffer sizes caused incorrect output.  If it's a problem, it can be reverted and a different solution can be found.

There are also other possibilities opened up now that viewports are usable, mainly the possibility of allowing emulated display modes (i.e. an envvar to force a specific fullscreen resolution).

### CODE REVIEW

**Comment by flibitijibibo** [comment]
I'll look at GNOME and Plasma tomorrow. At a glance this looks good on the Wayland front but the renderer details may be interesting for the other maintainers...

**Comment by sulix** [comment]
I haven't tested with a fractional scale factor yet, but this has a pretty significant regression under Plasma with a 2× scale factor. Bascially, non-high-DPI fullscreen appears to be completely broken.

Thus far, I've found that this results in mouse input broken in [Eliza](https://www.zachtronics.com/eliza/), and broken rendering in [Super Hexagon](https://superhexagon.com/).

Notably, Super Hexagon seems to have HIGHDPI support disabled, and shows the following in its log when run without this patch:
```
[notice] windowResized(1600, 900)
```

With this patch, the game renders into a quarter of the screen (with some random bleed-over to the rest of it), and logs:
```
[notice] windowResized(3200, 1800)
[notice] windowResized(3200, 1800)
```

A quick squizz at the code suggests that `native_{width,height}` would need to be scaled back down when used on a low-DPI window: multiplying it (almost) everywhere by `wind->scale_factor / driverdata->scale_factor` seems to do the trick, though probably is wrong for non-integer scaled stuff (I haven't read over the xdg-output stuff yet.) Similarly, just disabling this for windows without `SDL_WINDOW_ALLOW_HIGHDPI` would work, but not result in the correct scaling.

And, at the risk of asking awkward questions, is using wp-viewporter supposed to be the "correct" way of handling fractional scaling, or just a convenient hack until there's more proper support? If not, how worthwhile is our supporting it? If so, do we want to refactor things so that the integer-scaling support is a special-case of this (rely less heavily on `wind->scale_factor`)?

**Comment by flibitijibibo** [comment]
From what I read in the spec it does seem like a good way to get a non-integer scale out of the system, but I do wonder if it's possible to reuse the scale_factor more... but this extension does seem to be for both scaling and cropping, so at least when I look at it we're using it for one of the two cases. Could be wrong though!

My test system is 2x DPI so I'll see if I can't get this to cooperate today. If anyone wants another data point, Super Hexagon has a high-DPI mode in a Steam branch, password is "andkerplooey".

**Comment by sulix** [comment]
Quickly tested this against Super Hexagon and Eliza again, and both seem fixed. (I haven't had a chance to try the high-DPI Super Hexagon build yet: just the ancient non-steam one on my DPI-scaled laptop.)

I guess what I was trying to ask re: if this extension is the correct way of handling fractional scaling is related to what the story for windowed mode. With wp-viewporter, we're effectively inferring a fractional scale via the difference between the physical and logical resolution of the fullscreen display. But, if we're not in that fullscreen, dpi-aware happy path then we're effectively falling back to the integer scaling implementation. It seems a little weird to have support for more scaling factors in fullscreen than in windowed mode (though I guess we got that with physical resolution changes back in the day).

So, if we want to one day have fractional scaling for windowed mode, we're either going to need to wait for a new extension, or implement it on top of wp-viewporter outselves using that physical/logical size ratio of the display. If such a new extension appears, would we want to replace wp-viewporter completely, or would it still make more sense to use wp-viewporter for fullscreen, and something new for windowed mode?

This is all probably beyond the scope of this change, though: getting this working for fullscreen first should be okay. I think I'll probably get my head around it better once I've played with fractional scaling some more, too…

I'll see if I can try this out further on some games which actually try proper SDL_WINDOW_FULLSCREEN resolution changes, rather than always using FULLSCREEN_DESKTOP, too. Alas, I don't think there many of those around which use SDL2…

**Comment by Kontrabant** [comment]
It's entirely possible to do non-fullscreen windows that way, and I believe that Firefox does do this, but that enters a world of rounding pain and off-by-one errors, so that would best be addressed separately.  I wouldn't hold my breath on any fractional scaling extension.  It's been requested many times in recent years and nothing has materialized.  Either way, that's something that can be addressed later.  Fullscreen is the most important, as with fractionally scaled desktops, it can mean the difference between rendering to a 4k backbuffer with 8 million pixels, or an oversized 6k backbuffer with 21 million pixels.

Viewport support does open up many possibilities though, as you can hypothetically use any resolution you want.  The Wayland backend never has, and still doesn't, support mode changes, so calling SDL_SetWindowDisplayMode() has no effect.  With viewport support though, it's a possibility.

Added mouse pointer coordinate correction in the last push.  Pointer coordinates need to be mapped to the viewport dimensions instead of being truncated to the logical desktop space when we are basically faking a mode switch.

**Comment by flibitijibibo** [comment]
This got hit by https://github.com/libsdl-org/SDL/commit/40417b188a5dbfdd04297204caa77a3dabe8d7fd, which probably breaks the fractional scaling due to the scale comparison.

**Comment by Kontrabant** [comment]
> This got hit by [40417b1](https://github.com/libsdl-org/SDL/commit/40417b188a5dbfdd04297204caa77a3dabe8d7fd), which probably breaks the fractional scaling due to the scale comparison.

Yep, with fractional scaling the xdg-output dimensions divided by the native display resolution will never match the integer scale.  #ifdef-ed out for now in this branch.  Do all version of GNOME have this bug?  Never hit it personally.

**Comment by flibitijibibo** [comment]
I tested on 41.3 and 41.5. I couldn't find any reports for the issue so I'm not totally sure why it's like this...

**Comment by flibitijibibo** [comment]
Relaxed the requirement to narrow down special treatment to the one exact case (xdg isn't scaled, but wl is): https://github.com/libsdl-org/SDL/commit/713a675401cea4cde9f2bfb3771f3c32554bb278

**Comment by sulix** [comment]
Tried this with a 1.50× scaling factor under Plasma. I think it's on the whole pretty good, but may expose some DPI bugs in applications or compositors, as well as an issue with xdg-output's new scaled mode.

- I wasn't able to find any applications which support both SDL_WINDOW_FULLSCREEN (non-desktop) and High DPI displays. Hacking things together using sdl12-compat (with OpenGL scaling disabled) and a few titles, it looks like this is fixing things — the full, unscaled resolution is now supported correctly — and the xdg-output stuff broke things (the new "scaled" output mode is not properly scaled, so can't be used with SDL_WINDOW_FULLSCREEN).
- Eliza works, but renders a pixel of magenta at the top and bottom of the screen. It does that without this patch, though, anyway. As a non High-DPI-aware application, it's behaviour hasn't changed.


Of these, the big problem is actually with SDL_WINDOW_FULLSCREEN and non-HighDPI aware apps. Because we rely on the compositor for scaling here, it only works for the native resolution (and multiples thereof). So, we'd need to either get rid of the fractional scaled resolutions from the xdg-output patch (at least when HIGHDPI is disabled, but since that's a window property, not a display property, it's difficult to do), or support fractional scaling for non-HighDPI fullscreen windows, but with the correct scale, which should be possible with wp-viewporter, I'd think.
(As a bonus, that should let us "fake" arbitrary resolution changes in SDL fullscreen windows, à la XWayland, which would be a nice feature.)

tl;dr: Fractional scaling breaks quite a few things, quite a few of which are either the application's fault, the compositor's fault, or both. This patch seems like a strict improvement, but it doesn't fix everything, and the xdg-output patch exposes some more broken resolutions for non-High-DPI-aware applications. My personal feeling is that the "safe" option is to apply this, but disable the "correctly scaled" resolution from xdg-output, and that the "correct" solution is to extend this to implement wp-viewport-based scaling for non-HighDPI-aware applications.

**Comment by Kontrabant** [comment]
I'm currently working on extending this to support general mode switching emulation, so non-DPI aware applications should be able to pick a video mode, set it to fullscreen and have it behave as though it were native.  That should help a lot of older applications that just expect mode switching to work.

**Comment by flibitijibibo** [comment]
Since this possibly interacts with 12-compat, CCing @icculus as well - if there's anything specific we should test with this we'll give it a try.

**Comment by Kontrabant** [comment]
This last push extends viewport functionality to general mode emulation and, for now at least, exposes emulated display modes to test it with.

**Comment by sulix** [comment]
Awesome — this looks good to me.

There are definitely still some issues when the aspect ratio of the emulated mode is not the same as that of the real mode — forcing a game to 1024×768 or similar on a 16:9 screen definitely broke things. Given that we don't expose any such modes, though, it should be fine. (Though one day, emulating common 4:3 modes might be useful.)

sdl12-compat does have some existing code for [fake resolutions](https://github.com/libsdl-org/sdl12-compat/blob/main/src/SDL12_compat.c#L1771) and [calculating a letterboxing/pillarboxing viewport](https://github.com/libsdl-org/sdl12-compat/blob/main/src/SDL12_compat.c#L2396), which might be useful to crib from in a follow-up patch to add support for different aspect ratios.

There shouldn't need to be any special interaction with sdl12-compat. By default, it'll use fullscreen desktop, without HighDPI, and scale things itself using either ``SDL_SetRenderLogicalSize()`` or OpenGL FBOs. It can be overridden to actually change the resolution for OpenGL apps with the ``SDL12COMPAT_OPENGL_SCALING=0`` environment variable, and to use HighDPI with ``SDL12COMPAT_HIGHDPI=1``. Those are the options I've been using to test this out, and I haven't hit any issues with the current build, save for anything which hardcodes an unsupported resolution. (And, in that case, it should work unless people are force-disabling the sdl12-compat OpenGL scaling, and it was broken before.)

Eliza still gives magenta pixels, too, but that's true both with and without this patch.

So this is a strict improvement on what we have for everything I've tested so far, under KWin with 2× and 1.5× scaling.

File: src/video/wayland/SDL_waylandvideo.c [hunk 1]
```diff
@@ -260,6 +261,7 @@ Wayland_CreateDevice(int devindex)
     device->SetWindowHitTest = Wayland_SetWindowHitTest;
     device->FlashWindow = Wayland_FlashWindow;
     device->HasScreenKeyboardSupport = Wayland_HasScreenKeyboardSupport;
+    //device->GetWindowDrawableSize = Wayland_GetDrawableSize;
```
Comment (flibitijibibo) [inline src/video/wayland/SDL_waylandvideo.c:264]: Extra line
```suggestion
```

File: src/video/wayland/SDL_waylandvideo.c [hunk 2]
```diff
@@ -502,19 +575,52 @@ display_handle_done(void *data,
...
+    }
+    desktop_mode.refresh_rate = (int)SDL_round(driverdata->refresh / 1000.0); /* mHz to Hz */
+    desktop_mode.driverdata = driverdata->output;
+
+    /*
+     * The native display mode is only exposed separately from the desktop size if:
+     * the desktop is scaled and the wp_viewporter protocol is supported.
+     */
+    if (driverdata->scale_factor > 1.0f && video->viewporter != NULL) {
+
```
Comment (flibitijibibo) [inline src/video/wayland/SDL_waylandvideo.c:618]:
```suggestion
```

File: src/video/wayland/SDL_waylandvideo.c [hunk 3]
```diff
@@ -321,6 +325,7 @@ xdg_output_handle_logical_size(void *data, struct zxdg_output_v1 *xdg_output,
             return;
         }
     }
+
```
Comment (flibitijibibo) [inline src/video/wayland/SDL_waylandvideo.c:328]:
```suggestion
```

**Review by flibitijibibo — COMMENTED** [review]
Minor nits...

As far as the big picture goes I'm not a huge fan of hardcoded resolution arrays, but I don't know if there has even been a better way to make a resolution list programmatically.

Going to loop in Sam now since this touches core SDL files, but codewise this is fine. Will spend Monday running this through all the engines I have, and will merge into 22 if all goes well including Sam's review.

**Comment by Kontrabant** [comment]
Yeah, I'm not crazy about the resolution lists either, but there really doesn't seem to be a better way. XWayland resorts to the same thing.  

As far as non-matching aspect ratios, the Wayland viewport protocol doesn't let you position the viewport for proper letterboxing/pillarboxing, so exposing resolutions that aren't of the same aspect ratio, while possible, doesn't work well (maybe [GNOME 2338](https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/2338) seems to aim to address this, but I don't think we can globally count on compositors to do this).

**Comment by sulix** [comment]
Maybe we could have a much smaller resolution list which just has the few resolutions legacy games are likely to require (1080p, 1024×768, etc…)? Though if we're having a resolution list, we might as well put a decent selection in. Equally, we could get away with not adding any fake resolutions at this point: just having the fractional-scaling version should be enough for everything but specific legacy apps which need a specific resolution, and check that it's in the list instead of just setting it blindly. (And such programs were broken before, so this wouldn't be a regression).

For aspect-ratio letterboxing, could we work around the issues by using subsurfaces? i.e., have the game render to a (viewported) subsurface, which is positioned relative to an empty main surface?

Otherwise, I'm not too concerned if we don't support arbitrary resolutions here at the moment, given that the native resolution (and the appropriately scaled versions of it) work, very few SDL2 games actually want to change resolution (preferring SDL_WINDOW_FULLSCREEN_DESKTOP), and sdl12-compat will do its own resolution scaling, with working letterboxing, by default anyway.

**Comment by flibitijibibo** [comment]
Letterboxing and fancier aspect ratio work can come later I think... but for now let's cite Xwayland for the hardcoded resolutions and leave those as-is.

**Review by flibitijibibo — APPROVED** [review]
Tested latest with the following engines and all was well on the lab boxes:

- Source 1 (Portal 2)
- FNA (a bunch of stuff)
- Unity (Superliminal and a secret upcoming release)
- Custom SDL_Render stuff (VVVVVV, Eversion)
- Custom GL stuff (Wonder Boy (someone remind me to add high-DPI support to the Linux build))
- The Jackbox Party Pack

@slouken, if there are no objections to the SDL_video.c changes this can be applied!

File: src/video/SDL_sysvideo.h [hunk 4]
```diff
@@ -345,6 +345,7 @@ struct SDL_VideoDevice
     Uint32 next_object_id;
     char *clipboard_text;
     SDL_bool setting_display_mode;
+    SDL_bool no_display_mode_switching;
```
Comment (slouken) [inline src/video/SDL_sysvideo.h:348]: Where is this variable set to FALSE?
Also, please name it in a positive way for easier reading, e.g. allow_display_mode_switching or enable_display_mode_switching

**Comment by Kontrabant** [comment]
Quick fix for the emulated mode addition code so it adjusts for display orientation and better handles ultra-wide monitors.

I really wish there was a better way to handle the fullscreen desktop case as I'd wager that most applications just assume that the desktop resolution is the display's native resolution.  We do return the native buffer size when the DPI-aware flag is set, but for legacy applications, they end up drawing to a scaled buffer unless they use the real fullscreen flag, and then they have to make sure to use mode 0.

One small cosmetic opinion question regarding SDL_video.c, is "no_display_mode_switching" is right name for the new flag?  It needs to communicate that, while display modes can be set for the window, it shouldn't try to actually change the display mode.  For the record, this flag is needed as I tried just stubbing out the mode switch function in the Wayland backend, but when SDL thinks the actual mode switch has occurred, it does things like auto-minimize the window when focus is lost to prevent issues with it when switching the display mode back.

File: src/video/SDL_video.c [hunk 5]
```diff
@@ -1339,14 +1339,17 @@ SDL_UpdateFullscreenMode(SDL_Window * window, SDL_bool fullscreen)
...
-                /* only do the mode change if we want exclusive fullscreen */
-                if ((window->flags & SDL_WINDOW_FULLSCREEN_DESKTOP) != SDL_WINDOW_FULLSCREEN_DESKTOP) {
-                    if (SDL_SetDisplayModeForDisplay(display, &fullscreen_mode) < 0) {
-                        return -1;
-                    }
-                } else {
-                    if (SDL_SetDisplayModeForDisplay(display, NULL) < 0) {
-                        return -1;
+                /* Don't try to change the display mode if the driver emulates display changes. */
+                if (_this->no_display_mode_switching == SDL_FALSE) {
```
Comment (slouken) [inline src/video/SDL_video.c:1343]:
```c
if (!_this->disable_display_mode_switching)
```

**Comment by slouken** [comment]
@flibitijibibo, feel free to merge if you think it's ready for 2.0.22.
